File: python/fft.py
# -*- coding: utf-8 -*-
####################################################################################################
# Fast Fourier Transform -- Python 3 Version
# This version implements Cooley-Tukey algorithm for powers of 2 only.
#
# José Alexandre Nalon
####################################################################################################
# Since Python is an interpreted language, all you have to do is to invoque the interpreter to run
# this program:
#
# $ python fft_pure.py
#
# Notice that, since this implementation uses only standard Python 3 objects, you can try to run
# this script using Pypy 3, Cython 3 or any other Python 3 implementation.


####################################################################################################
# Import needed modules:
import fft_list
import fft_array
import fft_numpy
import numpy.fft as fft
from time import perf_counter                  # Time events;


####################################################################################################
# Definitions:
REPEAT = 50                                    # Number of executions to compute average time;


####################################################################################################
# Auxiliary Function:
def time_it(f, size, repeat=REPEAT):
    """
    Measure execution time through repeated calls to a (Fast) Fourier Transform function.

    :Parameters:
      f
        Function to be called;
      size
        Power of two of the size of the vector on which the transform will be applied;
      repeat
        Number of times the function will be called. Defaults to REPEAT.

    :Returns:
      The average execution time for that function with a vector of the given size.
    """
    x = [ j+0j for j in range(size) ]          # Generate a vector;
    t0 = perf_counter()                        # Start a timer;
    for j in range(0, repeat):                 # Repeated calls;
        f(x)
    return (perf_counter() - t0) / repeat      # Compute average;


####################################################################################################
# Main program:
if __name__ == "__main__":

    # Start by printing the table with time comparisons:
    print("+---------"*16 + "+")
    print("|    N    |   N^2   | N logN  "
          "| Direct  | CList   | Array   | NumPy   "
          "| Simpl.  | Matrix  "
          "| Recurs. | RArray  | RNumpy  "
          "| Itera.  | IArray  | INumpy  | Intern. |")
    print("+---------"*16 + "+")

    # Try it with vectors with size ranging from 32 to 1024 samples:
    for r in range(5, 11):

        # Compute the average execution time:
        n = 2**r
        dtime  = time_it(fft_list.direct_ft, n, REPEAT)
        ctime  = time_it(fft_list.lc_dft, n, REPEAT)
        atime  = time_it(fft_array.direct_ft, n, REPEAT)
        ntime  = time_it(fft_numpy.direct_ft, n, REPEAT)
        stime  = time_it(fft_numpy.simplified_ft, n, REPEAT)
        mtime  = time_it(fft_numpy.matrix_ft, n, REPEAT)
        rtime  = time_it(fft_list.recursive_fft, n, REPEAT)
        artime = time_it(fft_array.recursive_fft, n, REPEAT)
        nrtime = time_it(fft_numpy.recursive_fft, n, REPEAT)
        itime  = time_it(fft_list.iterative_fft, n, REPEAT)
        aitime = time_it(fft_array.iterative_fft, n, REPEAT)
        nitime = time_it(fft_numpy.iterative_fft, n, REPEAT)
        intime = time_it(fft.fft, n, REPEAT)

        # Print the results:
        print(f'| {n:7} | {n**2:7} | {r*n:7} '
              f'| {dtime:7.4f} | {ctime:7.4f} | {atime:7.4f} | {ntime:7.4f} '
              f'| {stime:7.4f} | {mtime:7.4f} '
              f'| {rtime:7.4f} | {artime:7.4f} | {nrtime:7.4f} '
              f'| {itime:7.4f} | {aitime:7.4f} | {nitime:7.4f} | {intime:7.4f} |')

    print("+---------"*16 + "+")

File: julia/fft.jl
####################################################################################################
# Fast Fourier Transform -- Julia Version
# This version implements Cooley-Tukey algorithm for powers of 2 only.
#
# José Alexandre Nalon
####################################################################################################
# Since Julia is an interpreted language, all you have to do is to invoque the interpreter to run
# this program:
#
# $ julia fft.jl


####################################################################################################
# Import needed modules:
import Printf                                  # Print messages;
import FFTW                                    # Internal Fast Fourier Transform, for comparison;


####################################################################################################
# Definitions:
const REPEAT = 500                             # Number of executions to compute average time;


####################################################################################################
# Auxiliary Function:
"""
    time_it(f, size, repeat=REPEAT)

Average execution time of a function.

This function calls a Fast Fourier Transform function `f` on a vector of `size` elements a certain
number of times (given by `repeat`), and returns the average execution time of the function.
"""
function time_it(f, size, repeat=REPEAT)
    x = Array{Float64}(0:size-1)               # Generate a vector;
    for j = 1:REPEAT                           # Repeated calls;
        f(x)
    end
end


####################################################################################################
"""
    direct_ft(x)

Fourier Transform of a vector `x` by direct definition.

Compute the Discrete Fourier Transform directly from the definition, an algorithm that has O(N^2)
complexity. The function returns a complex-number vector of the same size, with the coefficients of
the DFT.
"""
function direct_ft(x)
    N = length(x)                              # Length of the vector;
    X = zeros(Complex, size(x))                # Accumulates the results;
    W = exp(-2im*pi/N)                         # Twiddle factors;
    Wk = 1.0
    for k = 1:N                                # Compute the kth coefficient;
        Wkn = 1
        for n = 1:N                            #   Operates the summation;
            X[k] = X[k] + x[n]*Wkn             #     Computes every term;
            Wkn = Wkn * Wk                     # Update twiddle factors;
        end
        Wk = Wk * W
    end
    return X
end


####################################################################################################
"""
    recursive_fft(x)

Fast Fourier Transform of a vector `x` with a recursive decimation-in-time algorithm.

Compute the Fast Fourier Transform using a recursive decimation in time algorithm. This has
O(N log_2(N)) complexity. The function returns a complex-number vector of the same size, with the
coefficients of the DFT.
"""
function recursive_fft(x)
    if length(x) == 1                                  # A length-1 vector is its own FT;
        return x
    else
        N = length(x)                                  # Length of the vector;
        N2 = round(Int, N/2)
        Xe = recursive_fft(x[1:2:end])                 # Transform of even samples;
        Xo = recursive_fft(x[2:2:end])                 # Transform of odd samples;
        W = exp.(-2im*pi*Array(0:N2-1)/N)              # Twiddle factors;
        WXo = W .* Xo                                  # Repeated computation;
        X = [ Xe + WXo; Xe-WXo ]                       # Recombine results;
        return X
    end
end


####################################################################################################
"""
    bit_reverse(r, k)

Bit-reversed version of an integer number `k` with respect to `r` bits.
"""
function bitreverse(k, r)
    l = 0                                      # Accumulate the results;
    for i = 1:r                                # Loop on every bit;
        l = (l << 1) + (k & 1)                 # Test less signficant bit and add;
        k = k >> 1                             # Test next bit;
    end
    return l
end

"""
    iterative_fft(x)

Fast Fourier Transform of a vector `x` with an iterative decimation-in-time algorithm.

Compute the Fast Fourier Transform using an iterative in-place decimation in time algorithm. This
has O(N log_2(N)) complexity, and since there are less function calls, it will probably be
marginally faster than the recursive versions. The function returns a complex-number vector of the
same size, with the coefficients of the DFT.
"""
function iterative_fft(x)

    N = length(x)                              # Length of vector;
    r = round(Int, log(N)/log(2))              # Number of bits;
    X = complex(x)                             # Accumulate the results;
    for k = 0:N-1
        l = bitreverse(k, r)                   # Reorder the vector according to the
        X[l+1] = x[k+1]                        #   bit-reversed order;
    end

    step = 1                                   # Auxililary for computation of twiddle factors;
    for k = 1:r
        for l = 0:2*step:N-1
            W = exp(-1im * pi / step)          # Twiddle factors;
            Wkn = 1.0
            for n = 1:step
                p = l + n
                q = p + step
                X[q] = X[p] - Wkn*X[q]         # Recombine results;
                X[p] = 2*X[p] - X[q]
                Wkn = Wkn * W                  # Update twiddle factors;
            end
        end
        step = 2*step
    end
    return X

end


####################################################################################################
# Programa Principal
function main()

    # Start by printing the table with time comparisons:
    println("+---------+---------+---------+---------+---------+---------+---------+")
    println("|    N    |   N^2   | N logN  | Direta  | Recurs. | Itera.  | Interna |")
    println("+---------+---------+---------+---------+---------+---------+---------+")

    # Try it with vectors with size ranging from 32 to 1024 samples:
    for r = 5:10

        # Compute the average execution time:
        n = 2^r
        dtime = @elapsed time_it(direct_ft, n, REPEAT)
        rtime = @elapsed time_it(recursive_fft, n, REPEAT)
        itime = @elapsed time_it(iterative_fft, n, REPEAT)
        intime = @elapsed time_it(FFTW.fft, n, REPEAT)
        n = 2^r

        # Print the results:
        Printf.@printf("| %7d | %7d | %7d | %7.4f | %7.4f | %7.4f | %7.4f |\n",
                        n, n^2, r*n, dtime, rtime, itime, intime)
    end

    println("+---------+---------+---------+---------+---------+---------+---------+")
end

main()
